design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,438 @@
+          keyCephVersion: "20.2.0" # e.g.
+    ```
+
+### CephRbdMirror example
+CephRbdMirror keys have both local and non-local keys. For local keys, rotation follows the
+    methods explained for CephFilesystem above.
+
+For the non-local (peer) key:
+
+1. Reconciliation reads `spec.security.cephx.peer` configs to determine if reconcile should rotate.
+2. When key rotation is indicated, the peer key is rotated.
+3. Once reconciliation is nearing completion, the status is updated.
+
+Example showing input spec and output status, for clarity:
+
+```yaml
+# ...
+spec:
+  # ...
+  security:
+    cephx:
+      peer:
+        keyRotationPolicy: KeyGeneration
+        keyGeneration: 2
+  # ...
+status:
+  # ...
+  cephx:
+    local:
+      keyGeneration: 3
+      keyCephVersion: "20.2.2"
+    peer:
+      # before rotation
+      keyGeneration: 1
+      keyCephVersion: "20.2.0"
+      ## after rotation
+      # keyGeneration: 2
+      # keyCephVersion: "20.2.2"
+```
+
+### CephClient example
+
+CephClient stands out due to its simplicity. It will have one and exactly one key which can serve
+any purpose (it is not local nor peer). The interfaces are simplified to reflect this.
+
+```yaml
+kind: CephClient
+# ...
+spec:
+  # ...
+  security:
+    cephx:
+      keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+      keyGeneration: <int> # used with KeyGeneration
+    # (room for future spec.security options)
+  # ...
+status:
+  cephx:
+    keyGeneration: 1 # e.g.
+    keyCephVersion: "20.2.0" # e.g.
+```
+
+## Dependencies
+
+Ceph Tentacle (v20) provides a new `ceph auth rotate` command merged here:
+https://github.com/ceph/ceph/pull/58121. Rook will rely on this command to rotate keys.
+
+A few CephX technical details that are important for understanding CephX key rotation development
+are summarized below. More details here: https://docs.ceph.com/en/latest/dev/cephx/
+
+CephX keys minted by Rook are only used by Ceph for initial daemon connection. Internally, all Ceph
+connections use "service" keys with laddered expiration times.
+
+By default, Ceph service keys allow at least 2 hours from key rotation until the client must be
+updated with the new key. There are 3 service keys with TTLs 1 hour, 2 hours, and 3 hours from the
+time when service keys were last refreshed. Assuming the the first TTL expiration is imminent, it is
+still at least 2 hours until the 3-hour TTL expiration.
+
+Ceph does not allow keyrings to contain multiple keys for a given client/daemon. When a key is
+rotated, the old key is removed and new key replaces it with no ability to have a 2-valid-keys
+transition period.
+
+Ceph's `auth_service_ticket_ttl` and `auth_mon_ticket_ttl` config options allow users to
+shorten/lengthen that time as desired/needed. Note that this isn't immediate: it will take an
+unknown amount of time for internal service keys to update.
+
+Ceph config `debug_auth=30` provides maximum CephX debug logs, as a development aid.
```

Comment:
@BlaineEXE Do you also have plan to cover external ceph cluster as well? or its going to be a separate effort? and what is the Impact on the RDR if we change the keys?

Reply:
Make sense to document this one as they need to rotate the keys